ERCS/erc-7504.md

```diff
@@ -0,0 +1,402 @@
+
+### Router
+
+By ‘Router’, we mean a contract implementing the `Router` interface.
+
+A Router is a proxy contract. Its main, and only job is to route incoming function calls to the right implementation contract.
+
+The fallback function is triggered on a call to a non fixed function. The fallback function uses the first four bytes of the calldata (`msg.sig`) along with the `getImplementationForFunction` function to determine which implementation contract it should perform a delegate call on.
+
+This proposal specifies `Router` and `RouterState` as separate interfaces for the separation of concerns.
+
+The `Router` interface is only concerned with the core job of routing an incoming call to the appropriate implementation address via `Router.getImplementationForFunction`.
+
+The `RouterState` interface is concerned with storing and exposing the underlying data on which the return value of `Router.getImplementationForFunction` is based i.e. the ABI of the given dynamic contract.
+
+A Router contract can be deployed and used directly, without putting it behind a proxy. A Router contract can also be written with the intent to use it behind proxy contracts, so that multiple proxy contracts can route calls in the way defined in the same Router.
+
+This proposal makes no recommendations in this regard. Either form of usage maintains the usefulness of the standard.
+
+### RouterState, Extensions and building an ABI
+
+The `RouterState` interface is at the core of making dynamic contracts client friendly.
+
+Building a client — graphical or programatic — to interact with a smart contract requires an ABI, and a well defined interpretation of the ABI to instruct the interaction.
+
+Dynamic contracts pose two immediate challenges in this regard:
+
+1. The callable functions of a dynamic contract may change. This poses a challenge for clients, since they cannot always anticipate the same ABI to work for a dynamic contract.
+2. The callable functions of a dynamic contract are spread across many implementation contracts.
+3. Such changes to a contract's ABI may warrant the expression of changes in expected usage and behaviour of the smart contract.
+
+The `RouterState` interface creates a natural grouping that we call `Extension`. At any given point, a dynamic contract knows of a fixed set of extensions.
+
+An `Extension` is an abstraction of an implementation contract used by a dynamic contract.
+
+| metadata.name           | The name of an Extension.                                             |
+| ----------------------- | --------------------------------------------------------------------- |
+| metadata.metadataURI    | A URI where we store some information about the Extension.            |
+| metadata.implementation | The implementation contract address of the Extension.                 |
+| metadata.functions      | All functions implemented in the Extension’s implementation contract. |
```

Comment:
This table seems like it belongs more in the specification section, though I could be wrong. It doesn't seem to be _explaining_ why you made a choice, but rather explaining how implementers should use these fields.